### PR TITLE
ZEPPELIN-830 Improve table display to handle large data

### PR DESCRIPTION
**Pull request by r-kamath**
### What is this PR for?
This is an improvement for table display. By using [Handsontable](https://github.com/handsontable/handsontable) we can load and display more data in paragraphs. Tested using sample tabular data consisting 10000x80 cells

### What type of PR is it?
Improvement

### Todos
* [x] - Decimal formatting
* [x] - Rendering without angular/html directives
* [x] - UI fixes
* [x] - License doc update

### What is the Jira issue?
* https://issues.apache.org/jira/browse/ZEPPELIN-830

### How should this be tested?
TODO : test cases to follow

### Screenshots (if appropriate)
apr 26 2016 17 38

### Questions:
* Does the licenses files need update? Yes
* Is there breaking changes for older versions? No
* Does this needs documentation? Maybe

### CODE REVIEW

**Comment by prabhjyotsingh** [comment]
@r-kamath build fails with 


    Failed tests: 
      SparkParagraphIT.testSqlSpark:169 Paragraph from SparkParagraphIT of testSqlSpark result: 
    Expected: "age job marital education balance\n30 unemployed married primary 1,787"
         but: was "age\njob\nmarital\neducation\nbalance\n30 unemployed married primary 1787\nage\njob\nmarital\neducation\nbalance"

Looks like this require selenium test update.

**Comment by astroshim** [comment]
I passed the build with no problems and functionality is working very well. 
but @r-kamath you should make the green CI.
Thanks.

**Comment by felixcheung** [comment]
How does this work with some of the pending data table changes?
like #6, #714, #725, #761

**Comment by r-kamath** [comment]
Won't affect #714 and #725

#6 and #761 are using jQuery Datatables
This PR is a proposal to use Handsontable instead of datatables. Handsontable is more like a spreadsheet and has good performance with many builtin [features](https://handsontable.com/features.html) and plugin support.

I've done a basic performance test with 20000x80 rows and columns data and few paragraphs with a subset.
PR #761 : https://drive.google.com/open?id=0B8L8gc9P-C-6eEhHeFVRd3cxNEE
PR #858 : https://drive.google.com/open?id=0B8L8gc9P-C-6dEo3OWVWa1J3QTQ

**Comment by felixcheung** [comment]
We would need export to file feature which seems like is not supported in the free Handsontable version?

**Comment by r-kamath** [comment]
Right, it is supported only in pro ver. Handsontable's Export feature is in front-end. IMHO export should be a backend feature and will be a good addition to our Notebook REST api.

File: zeppelin-web/src/index.html [hunk 1]
```diff
@@ -132,6 +134,10 @@
     <script src="bower_components/bootstrap3-dialog/dist/js/bootstrap-dialog.min.js"></script>
     <script src="bower_components/floatThead/dist/jquery.floatThead.js"></script>
     <script src="bower_components/floatThead/dist/jquery.floatThead.min.js"></script>
+    <script src="bower_components/zeroclipboard/dist/ZeroClipboard.js"></script>
+    <script src="bower_components/moment/moment.js"></script>
+    <script src="bower_components/pikaday/pikaday.js"></script>
+    <script src="bower_components/handsontable/dist/handsontable.js"></script>
```
Comment (Leemoonsoo) [inline zeppelin-web/src/index.html]: pickaday.js, moment.js, ZeroClipboard.js, pikaday.css are they required to use handsontable?

**Comment by Leemoonsoo** [comment]
Is it possible to handle '%html' inside of table cell?

if i run
```
println(s"""%table key\tvalue
html\t%html <dev style='color:blue'>blue</div>
""")
```

Master branch
image

This PR
image

**Comment by r-kamath** [comment]
@Leemoonsoo that's in the todo list. Can be done either in FE or in server. Thoughts?

File: zeppelin-web/src/index.html (hunk 1 above)
Comment (r-kamath) [inline zeppelin-web/src/index.html]: ZeroClipboard and pikaday.js are not required, need to see how to remove them from the build.
As of now moment.js is used only for number formatting (7bb508e)

**Comment by r-kamath** [comment]
@Leemoonsoo  `%html` rendering is fixed in 44ef274

File: zeppelin-web/src/index.html (hunk 1 above)
Comment (r-kamath) [inline zeppelin-web/src/index.html]: I was wrong. All three deps are required. [Use without pikaday #2841](https://github.com/handsontable/handsontable/issues/2841)

**Comment by Leemoonsoo** [comment]
%html rendering is working nicely! If pickaday.js, moment.js, ZeroClipboard.js are required to use handsontable, they also need to be in 'zeppelin-distribution/src/bin_license/LICENSE'

**Comment by r-kamath** [comment]
@Leemoonsoo license doc updated and rebased 2e422ea 4fbb22b

**Comment by prabhjyotsingh** [comment]
LGTM! nice improvement 👍

**Comment by Leemoonsoo** [comment]
@r-kamath Are there any more works left for unchecked todo item 'UI fixes', in the description of this PR?

**Comment by r-kamath** [comment]
@Leemoonsoo the only pending item is the sort header arrow style. Should we try something more stylish ?
screen shot 2016-05-04 at 11 44 08 pm

screen shot 2016-05-04 at 11 42 36 pm

**Comment by Leemoonsoo** [comment]
Arrow style looks fine. I think it's good to be merged!

**Comment by r-kamath** [comment]
@Leemoonsoo cool. I have checked the 'UI fixes'. thanks

**Comment by Leemoonsoo** [comment]
Last CI test failure is unrelated to the change. LGTM.

**Comment by felixcheung** [comment]
does this PR change the export to CSV capability?

**Comment by corneadoug** [comment]
@felixcheung No, There is still no export to CSV feature

**Comment by r-kamath** [comment]
@felixcheung this PR is only for improving table performance. CSV/TSV export is in #725 and #714

**Comment by prabhjyotsingh** [comment]
@r-kamath  can you resolve branch conflict.

**Comment by r-kamath** [comment]
@prabhjyotsingh conflict resolved f13ba91

**Comment by prabhjyotsingh** [comment]
Thanks @r-kamath for contributing towards front-end performance. 
Merging if no more discussion.

**Comment by corneadoug** [comment]
Anyway to not have:
* zeroclipboard
* pikaday
* moment

included when installing Handsontable?

**Comment by r-kamath** [comment]
@corneadoug  [Use without pikaday #2841](https://github.com/handsontable/handsontable/issues/2841)

**Comment by corneadoug** [comment]
@r-kamath https://github.com/handsontable/hot-builder
I understand that by default it seems to be needed, however I'm reticent to include 3 different dependencies that we are not using.

**Comment by r-kamath** [comment]
Not sure if we can avoid those dependencies by choosing hot-builder!
Though we are not using the copy/ paste, date picker and date formatting, it is still a feature in Handsontable.
IMO we shouldn't be worried about those three dependencies because it is not a part of our web bower.json

**Comment by prabhjyotsingh** [comment]
I would agree with @r-kamath, as long as those dependencies are not been written in bower.json we should be ok.
https://github.com/apache/incubator-zeppelin/pull/858/files#diff-6bcc7568e5b6af0b1499ab172fc66af4L33

**Comment by corneadoug** [comment]
I don't get why its not important to have additional dependencies that we are not using in our build if its not in bower.json, maybe somebody can explain more.

**Comment by r-kamath** [comment]
@corneadoug Again, this is an issue on handsontable please follow these links. Those entries in index.html should disappear when handsontable team fix the issue.
https://github.com/handsontable/handsontable/issues/2841
https://github.com/handsontable/handsontable/issues/2974

Also it is included in the [Improvement-suggestions](https://github.com/handsontable/handsontable/wiki/Improvement-suggestions) section.

More details here : https://github.com/handsontable/handsontable

@Leemoonsoo @prabhjyotsingh what say?

**Comment by corneadoug** [comment]
@r-kamath I don't want to make this PR wait forever.
If it was easy to do, I would prefer removing those dependencies. Mainly because everytime we want to add a dependency, we should ask ourselves if its really necessary, so adding 3 dependencies that are not being used is even worst.

Since Handsontable seems to be helpful in term of performances, it might still be worth it to have those unnecessary dependencies. However, after this PR, we should try to find a way to get rid of them somehow (by keeping an eye for a fix in handsontable, or making a PR there, or even trying their hotbuilder)

**Comment by r-kamath** [comment]
@corneadoug I'm following up on handsontable repo. Will find some time to explore for a potential PR.

**Comment by corneadoug** [comment]
Merging if there is no more discussions

**Comment by yantaiv** [comment]
Great Job. I have a silly question, how do you debug your code in zeppelin, I mean , what IDE do you use and how you make breakpoint or sth like that ...

**Comment by corneadoug** [comment]
@yantaiv Better ask that in the mailing list, you would get more answers
But overall:
You can use any IDE you want, there is informations about dev mode in the README

**Comment by gauravkumar37** [comment]
Thanks for the wonderful commit.
However, I faced one issue. I am not able to select any data in the table using mouse for copying it. (checked on FF & chrome latest versions)

**Comment by corneadoug** [comment]
@gauravkumar37 handled in #973

**Pull request closed without merging**